### PR TITLE
feat[platform]: add email, hyperlink data types to server side. 

### PR DESCRIPTION
**Pull request by mswertz**
Todo: user interfaces

- [x] backend + tests
- [x] schema editor (ability to select the type)
- [x] forms (client side validation)
- [x] data explorer (clickable links in the table/cards)
- [x] update documentation

### CODE REVIEW

File: backend/molgenis-emx2-exampledata/src/main/java/org/molgenis/emx2/examples/PetStoreExample.java [hunk 1]
```diff
@@ -56,7 +57,7 @@ public static void create(SchemaMetadata schema) {
         table(ORDER)
             .add(column(ORDER_ID).setPkey())
             .add(column("pet").setType(REF).setRefTable(PET))
-            .add(column(QUANTITY).setType(INT).setValidation("{quantity} >= 1"))
+            .add(column(QUANTITY).setType(LONG).setValidation("{quantity} >= 1"))
```
Comment (mswertz) [inline backend/molgenis-emx2-exampledata/src/main/java/org/molgenis/emx2/examples/PetStoreExample.java:60]: added as smoke test of new LONG type

File: backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/EvaluateExpressions.java [hunk 2]
```diff
@@ -68,6 +68,10 @@ public static Object calculateComputedExpression(String expression, Row row) {
 
   public static void checkValidation(Map<String, Object> values, Collection<Column> columns) {
     for (Column column : columns) {
+      // system based validation
+      column.getColumnType().validate(values.get(column.getName()));
```
Comment (mswertz) [inline backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/EvaluateExpressions.java:72]: this is for email and hyperlink

File: backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/graphql/GraphqlTableFieldFactory.java [hunk 3]
```diff
@@ -151,6 +151,15 @@ private GraphQLObjectType createTableObjectType(Table table) {
               GraphQLFieldDefinition.newFieldDefinition()
                   .name(id)
                   .type(GraphQLList.list(Scalars.GraphQLInt)));
+        case LONG:
+          tableBuilder.field(
+              GraphQLFieldDefinition.newFieldDefinition().name(id).type(Scalars.GraphQLBigInteger));
```
Comment (mswertz) [inline backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/graphql/GraphqlTableFieldFactory.java:156]: actually, graphql doesn't officially know about LONG/BigInteger so officially this is deprecated feature

File: apps/styleguide/src/tables/RenderValue.vue [hunk 4]
```diff
@@ -1,23 +1,37 @@
 <template>
-  <span v-if="row[col.name] && Array.isArray(row[col.name])">
-    <span v-for="(val, idx) in row[col.name]" :key="idx">
-      {{ col.refLabel ? applyJsTemplate(col.refLabel, val) : val }}
+  <span v-if="row[col.id] && Array.isArray(row[col.id])">
+    <span v-for="(val, idx) in row[col.id]" :key="idx">
+      <a v-if="col.columnType == 'HYPERLINK_ARRAY'" :href="val" target="_blank">
```
Comment (mswertz) [inline apps/styleguide/src/tables/RenderValue.vue:4]: can you assume val is valid href?

File: backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTypeUtils.java [hunk 5]
```diff
@@ -49,94 +51,58 @@ private SqlTypeUtils() {
...
-        return row.getDateTimeArray(name);
-      case JSONB:
-        return row.getJsonb(name);
-      case JSONB_ARRAY:
-        return row.getJsonbArray(name);
-      default:
-        throw new UnsupportedOperationException(
-            "Unsupported columnType found:" + c.getColumnType());
-    }
+    return switch (c.getPrimitiveColumnType()) {
```
Comment (mswertz) [inline backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTypeUtils.java:54]: nice :-)

**Review by mswertz — COMMENTED** [review]
code looks good. I committed a few minor changes. TODO is the functional review.

File: apps/styleguide/src/tables/RenderValue.vue (hunk 4 above)
Comment (chinook25) [reply apps/styleguide/src/tables/RenderValue.vue:4]: val has been validated on input, so it should be valid. 
I (we?) didn't know about that display, will change it :)

**Review by chinook25 — CHANGES_REQUESTED** [review]

**Review by chinook25 — APPROVED** [review]

**Review by mswertz — COMMENTED** [review]
We forgot LONG_ARRAY in frontend (I think it might be in backend already)

image

**Comment by mswertz** [comment]
Kreeg de volgende foutmelding toen ik dikke string copy/paste met max number

image

**Comment by mswertz** [comment]
When I add a number larger then max long it becomes negative number
image
becomes
image

**Comment by mswertz** [comment]
when I add a space I get invalid email. However, we normally would trim this value during save. Shall we also trim the validation?
image

**Comment by mswertz** [comment]
I think we should be more strict. I now added following example and obviously link to www.rug.nl failed.
image

I propose we do require http:// or https:// before

**Comment by chinook25** [comment]
I just realized we still need to do documentation

**Comment by MaxPostema** [comment]
> I think we should be more strict. I now added following example and obviously link to [www.rug.nl](http://www.rug.nl) failed. image
> 
> I propose we do require http:// or https:// before

- Note: that is is how it works for emx1 (www.rug.nl being a relative link)
- I think datamanagers use relative links to link in the applications itself or not?

**Pull request merged**